My Sites > Posts: update comment count when adding new comment

v18 opened the pull request:
This fixes #4889 by adding a new component `PostTotalComments` for displaying the comment icon in the post list for a site. When a new comment is added to the post, a new action type, `COMMENTS_COUNT_INCREMENT`, is dispatched to increment the comment count in `state.comments`.
# Here's what the fixed behaviour looks like:

![comments-updating](https://cloud.githubusercontent.com/assets/11487924/17855521/19205f0c-6847-11e6-9222-488142f5bc0b.gif)
# How to test
1. Go to the posts list for one of your sites, for example:
   http://calypso.localhost:3000/posts/my/mysite.wordpress.com
2. Find a post that has a comments icon
3. Click the comments icon & add a new comment

I tested these cases manually as well
- Jetpack comments
- wpcom comments
- starting at 0 and 1 

You can run the new batch of tests using
`npm run test-client client/my-sites/posts/test`
`npm run test-client client/state/comments/test/reducer.js`
# Reusing

This component can be re-used in other places in Calypso:
1. The reader full-post view for the feed, for example: https://wordpress.com/read/feeds/45339984/posts/1126008375
2. The reader full-post view for a blog at https://wordpress.com/read/blogs/$siteId/posts/$postId

If we re-use this component, it should probably be moved to `client/components`, or is there a better place?

Test live: https://calypso.live/?branch=fix/4889-posts-list-comment-count

mtias commented:
Nice! I think this should be donde directly in https://github.com/Automattic/wp-calypso/tree/master/client/blocks/comment-button, which right now just takes a `commentCount` prop. The component already has a `showLabel` prop that should match the usage here in `/posts`.

v18 commented:
@mtias thanks! Looks like I missed CommentButton before. I'll add that in & push my changes.

nb commented on client/my-sites/posts/post-total-comments.jsx:
```diff
@@ -0,0 +1,91 @@
+		);
+	},
+
+	getCommentCount: function( syncedCount = this.props.syncedCount, originalCount = this.props.originalCount ) {
```
Do we need the function arguments for something different than testing? What would it take to not have testing-specific code?

v18 replied:
I added the `getCommentCount` code more for readability purposes than testing. I can certainly move it out of the function and add a few tests that don't rely on the function specifically, but the state instead. Let me know if you think I should.

v18 replied:
Nvm, I misread your question earlier :) I added in the _arguments_ to the function from functional programming muscle memory, and it of course makes it easier to test too. Happy to get rid of the function entirely as I mentioned in my previous comment.

nb commented on client/state/comments/actions.js:
```diff
@@ -221,6 +222,13 @@ export function writeComment( commentText, siteId, postId, parentCommentId ) {
+				type: COMMENTS_COUNT_INCREMENT,
+				siteId,
+				postId,
+				incrementBy: 1
```
Are we sure that the comment count increased by one after our API call?

v18 replied:
In this case, yes. This dispatch is happening inside the `writeComment` function, which adds a new comment when it's added by a user in `reader/comments/form.jsx`.

nb replied:
What if somebody else added a comment in the meantime? What is our ultimate source of truth?

v18 replied:
If someone else added a comment in the meantime, that comment would not be displayed in the comment list and it would not count towards the comment total. If the page is refreshed, then our components would get and display that information.

This is a limitation of `PostCommentList` since it only fetches the full list of comments when loading for the first time, and does not fetch a new list when a new comment is added. 

The ultimate source of truth for the comment count is `state.comments.totalCommentsCount`. The comment list and comment total are not currently being kept in sync with the database after a new comment is made, but the comment count matches with the actual comments shown on the page.

In the future we could keep the state in sync with the database by either (1) pulling new data on every comment submission or (2) polling it for data periodically.

v18 commented:
Added new commits to address issues mentioned above.

## Overview of changes

- Use `CommentButton` instead of using new `PostTotalComments` in my-sites/posts, cc @mtias 
- Hook up `CommentButton` to state
- Add tests for `CommentButton`
- Update all `CommentButton`s to pass down postId and siteId as props - 4 instances used by the reader
- No longer using `getCommentCount()` helper function cc @nb 
- Fix bug for comments that were awaiting moderation
- Update docs for `CommentButton`

### Tests
You can run the unit tests using:
`npm run test-client client/state/comments/test/reducer.js` (same as before)
`npm run test-client client/blocks/comment-button/test` (new)

### Inadvertent small visual fix 

My fixes also fixed a small display issue on my-sites/posts on smaller screens:

Before:

<img width="549" alt="old-display" src="https://cloud.githubusercontent.com/assets/11487924/17901988/7aa95f98-6933-11e6-8d55-e17bab51733d.png">

After:

<img width="548" alt="new-display" src="https://cloud.githubusercontent.com/assets/11487924/17901995/81b5c290-6933-11e6-9912-3c6170698d78.png">


### Reader changes

There's one issue with the fixes: the reader seems to be undergoing some major changes right now, so it would probably be good to make sure these changes are communicated to that team before pushing to mater (even though they won't cause trouble).

nb commented on client/blocks/comment-button/index.jsx:
```diff
@@ -2,20 +2,26 @@
 		commentCount: React.PropTypes.number,
-		showLabel: React.PropTypes.bool
+		showLabel: React.PropTypes.bool,
+		postId: React.PropTypes.number.isRequired,
```
Why do we need `postId` and `siteId` here? We don’t seem to be using them in the pure comment button.

v18 replied:
I expect only `CommentButton` to be used in production, which does always require `postId` and `siteId`.

`PureCommentButton` should only be used in tests and dev docs. I'll add a comment noting this in my next commit.

mtias replied:
We don't really have this pattern elsewhere, and prefer to test the full component in devdocs. I think we can get rid of `PureCommentButton`.

nb commented on client/blocks/comment-button/index.jsx:
```diff
@@ -39,7 +45,7 @@ const CommentButton = React.createClass( {
 		let label;
 		const containerTag = this.props.tagName,
-			commentCount = this.props.commentCount;
+			commentCount = ( this.props.syncedCount === undefined ? this.props.commentCount : this.props.syncedCount );
```
Shouldn’t we include the `syncedCount` prop in the `propTypes`? Actually, why do we need both `commentCount` and `syncedCount`?

v18 replied:
> Shouldn’t we include the syncedCount prop in the propTypes?

Yes, we should :), that was an omission error on my part, and I'll add it in my next commit.

> Actually, why do we need both commentCount and syncedCount?

For now we need both for slightly complicated reasons:

- `commentCount` is the prop passed in when using CommentButton, and it's the default comment count we'll display
- `syncedCount` is the redux state's current comment count for the post. It can be undefined if the list of comments has not yet been fetched for the post, which is why we need a default.

The total comment count could be fetched separately when rendering CommentButton, but here's why I decided against it:

- we could fetch it using the API endpoint `/posts/$post_ID/replies`, but it's an extra network request
- we can't get it from comment redux state, because it's unavailable until we [fetch all comments](https://github.com/Automattic/wp-calypso/blob/master/client/state/comments/actions.js#L61)
- fetching all comments would also mean another network request
- we do know the comment count when displaying a post, but it's stored in flux, and can't neatly be added to redux's state
- therefore, to reduce # of network requests, we pass in the comment count we know from the post

In the future, when we're using redux to store post state, we can populate the redux `state.comments.totalCommentCount`, and remove the `commentCount` prop

mtias replied:
> In the future, when we're using redux to store post state, we can populate the redux state.comments.totalCommentCount, and remove the commentCount prop

We are using redux to hold posts state already, it's just not used yet for the main post-cards in `/posts` (but it is used for the draft items in the sidebar).

If you can't rely on one selector for getting the comment count, I'd probably rename the props to have an `initialCount` or `defaultCount` for passing the value down from the post. The data read from state should be the canonical, and thus match the main prop. (Which, by the way, could be just `count`.)

v18 replied:
> We are using redux to hold posts state already, it's just not used yet for the main post-cards in /posts (but it is used for the draft items in the sidebar).

You're right, sorry, I overgeneralized :)

As for the two count props, I answered that in [my main comment below](https://github.com/Automattic/wp-calypso/pull/7581#issuecomment-243971060).

nb commented on client/reader/full-post/main.jsx:
```diff
@@ -346,7 +346,7 @@ FullPostDialog = React.createClass( {
 
 			if ( shouldShowComments ) {
-				buttons.push( <CommentButton key="comment-button" commentCount={ this.props.commentCount } onClick={ this.handleCommentButtonClick } tagName="div" /> );
+				buttons.push( <CommentButton key="comment-button" commentCount={ this.props.commentCount } onClick={ this.handleCommentButtonClick } tagName="div" postId={ post.ID } siteId={ post.site_ID } /> );
```
Do we need to pass `commentCount` if it will be retrieved from the store via `getPostTotalCommentsCount()`?

v18 replied:
The result of `getPostTotalCommentsCount()` isn't always defined, so we need another way to get at the comment count. I explained the reason for passing in `commentCount` instead of fetching the total comments count in CommentButton in my [reply above](https://github.com/Automattic/wp-calypso/pull/7581#discussion_r76358978).

mtias commented on client/my-sites/posts/style.scss:
```diff
@@ -195,10 +195,28 @@
 			}
 			span {
-				display: inline-block;
+				// display: inline-block;
```
Is this supposed to be removed?

v18 replied:
Yes, removed in 80f851a.

mtias commented on client/blocks/comment-button/index.jsx:
```diff
@@ -68,4 +77,11 @@ const CommentButton = React.createClass( {
+	return { syncedCount };
+} )( CommentButton );
+
+export { CommentButton as PureCommentButton };
```
I mentioned it in another comment, but why do we need a pure comment button?

mtias commented on client/blocks/reader-full-post/index.jsx:
```diff
@@ -103,7 +103,9 @@ export class FullPostView extends React.Component {
 								onClick={ this.handleCommentClick }
-								tagName="div" />
+								tagName="div"
+								postId={ post.ID }
```
We prefer to use the global ID to identify posts.

mtias commented on client/my-sites/posts/post.jsx:
```diff
@@ -16,9 +16,11 @@ import PostHeader from './post-header';
 import utils from 'lib/posts/utils';
 import updatePostStatus from 'lib/mixins/update-post-status';
 import analytics from 'lib/analytics';
+import { shouldShowComments } from 'reader/comments/helper';
```
Is this logic applicable outside the Reader (and will it remain like that)? The helper function seems to disable comments for Jetpack sites, for instance.

v18 commented:
I added some new commits to address some of the issues in the line comments.

> We don't really have this pattern elsewhere ... I think we can get rid of PureCommentButton.

I removed the separate export of PureCommentButton in f30a87a.

> ... and prefer to test the full component in devdocs.

I haven't updated devdocs to use the full component yet since I've got a question for you, @mtias. I use `postId = 1` and `siteId = 1` in the devdocs. Right now this works since the component doesn't fetch any data, but in the future it probably will. In that case those values would result in a permissions error from the API. Do you think I should address it now by using different values, or leave it for later and use the full component in devdocs now?

> I'd probably rename the props to have an initialCount or defaultCount for passing the value down from the post. The data read from state should be the canonical, and thus match the main prop. (Which, by the way, could be just count.)

I updated CommentButton to use only `count` for both the initial prop passed in, and the comment count we get from the redux state in 0843806.

> Is this logic applicable outside the Reader (and will it remain like that)? The helper function seems to disable comments for Jetpack sites, for instance.

Good point. Even if the logic is applicable, you're right, we shouldn't rely on it to remain that way in the future. I refactored it to figure out whether to display the comment button internally in 0152592 and 09132f9.

> We prefer to use the global ID to identify posts.

I'm not sure what you mean by this. Can you elaborate please? :)

mtias commented:
> Do you think I should address it now by using different values, or leave it for later and use the full component in devdocs now?

Yes, otherwise it could cause some havoc in the future if someone doesn't check what site and post it's using. Look at a few other examples in devdocs/blocks, some use the primary site of the current user, others use example sites or `en.blog` if it's read only.

You could also use what the Reader blocks ("Search Cards") use as well.

> I'm not sure what you mean by this. Can you elaborate please? :)

Sure :) `postId` is local to a site, which means it can run into conflicts across sites and forces you to always pass siteId alongside it to select accurately. But posts also have a global id that doesn't require the site. See example usages in:

- https://github.com/Automattic/wp-calypso/blob/master/client/blocks/post-status/docs/example.jsx#L37
- https://github.com/Automattic/wp-calypso/blob/master/client/blocks/post-item/index.jsx#L58

aduth commented on client/blocks/comment-button/index.jsx:
```diff
@@ -38,34 +44,43 @@ const CommentButton = React.createClass( {
 			containerTag, {
-				className: 'comment-button',
+				className: classNames( {
+					'comment-button': true,
```
If a property of the object passed to `classNames` is always true, you can alternative pass it as a separate string argument, i.e.:

```js
className: classNames( 'comment-button', {
	'is-empty': count === 0
} )
```

aduth commented on client/blocks/comment-button/index.jsx:
```diff
@@ -38,34 +44,43 @@ const CommentButton = React.createClass( {
+export default connect( ( state, ownProps ) => {
+	const { siteId, postId } = ownProps,
+		count = getPostTotalCommentsCount( state, siteId, postId );
+	if(count === undefined) {
```
Needs more spacing on this line:

```
wp-calypso (fix/4889-posts-list-comment-count) $ ./node_modules/.bin/eslint client/blocks/comment-button/index.jsx 

/Users/andrew/Documents/Code/wp-calypso/client/blocks/comment-button/index.jsx
  82:2   warning  Expected space(s) after "if"             keyword-spacing
  82:4   warning  There must be a space inside this paren  space-in-parens
  82:24  warning  There must be a space inside this paren  space-in-parens

✖ 3 problems (0 errors, 3 warnings)
```

aduth commented on client/state/comments/actions.js:
```diff
@@ -221,6 +222,15 @@ export function writeComment( commentText, siteId, postId, parentCommentId ) {
 				skipSort
 			} );
 
+			if( comment.status === 'approved' ) {
```
Need space after `if`

```
wp-calypso (fix/4889-posts-list-comment-count) $ ./node_modules/.bin/eslint --quiet --rule "keyword-spacing:2" client/state/comments/actions.js 

/Users/andrew/Documents/Code/wp-calypso/client/state/comments/actions.js
  225:4  error  Expected space(s) after "if"  keyword-spacing

✖ 1 problem (1 error, 0 warnings)
```

aduth commented:
A post object also tracks comment count by `post.discussion.comment_count`, which could similarly watch for the `COMMENTS_COUNT_INCREMENT` action from [its relevant reducer](https://github.com/Automattic/wp-calypso/blob/7931f37f00a509f244655521f7ade67c270a2eb1/client/state/posts/reducer.js#L132-L230). That said, I'm not sure we're using that data anywhere currently, but it could be the source of some confusion if not kept in sync.

lancewillett commented:
Re-ping @v18 -- needs reply based on most recent feedback.

(This came up in a search for pull requests pending without an update in last 2 months.)

lancewillett commented:
Ping @v18 again. Should we close it?

v18 commented:
Hey @lancewillett - thanks for the ping! :) I'm afk right now, but will see about either closing this or finishing up the work (not much left to do) starting the week of Dec 19.

v18 commented:
Closing this for the time being.
